### PR TITLE
[Lang] Support TensorType for Dead-Store-Elimination

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: #

### <samp>🤖 Generated by Copilot at 07b919d</samp>

This pull request enhances the data source analysis and the load-store forwarding analysis for tensors and matrices, fixes a memory initialization bug for CUDA devices, and adds a test case for memory reuse. It modifies the files `data_source_analysis.cpp`, `control_flow_graph.cpp`, `analysis.h`, `cuda_device.cpp`, `compile_to_offloads.cpp`, and `test_memory.py`.

### <samp>🤖 Generated by Copilot at 07b919d</samp>

*  Fix a bug where uninitialized memory could cause incorrect results or crashes for CUDA devices by memseting the allocated memory to zero in `CUDADevice::allocate_memory` ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-7919f5d7e33aafc72f27ed93febc58a0ac77c220ae718bf78ef134dad3790654R63-R66))
*  Add and modify functions for data source analysis to handle aliasing issues and support load-store forwarding and dead store elimination for tensors and matrices ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-eb67594608637d4fc58874bf769d4a8eea22c9d100ada4767d913ad57671a980L13-R13), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-eb67594608637d4fc58874bf769d4a8eea22c9d100ada4767d913ad57671a980L24-R24), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-eb67594608637d4fc58874bf769d4a8eea22c9d100ada4767d913ad57671a980L37-R46), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-eb67594608637d4fc58874bf769d4a8eea22c9d100ada4767d913ad57671a980L58-R78), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-f6bc75768d2e24c782fefa45a7232d0e2b2bae091e697040e7f442a77d80ad45R113), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL118-R122))
*  Modify the control flow graph logic to use the data source analysis functions and check for tensor involvement and aliased addresses when performing store-to-load forwarding, identical store elimination, dead store elimination, and identical load elimination ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR128-R130), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL137-R154), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL146-R171), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR184-R186), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR230-R234), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL212-R247), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR253-R254), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR268-R290), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL259-R338), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL272-R355), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL318-R381), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR481-R520), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR527-R546), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL433-R554), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL494-R577), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL506-R589), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL520-R610), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL537-R632), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL546-R660), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL556-R676), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL574-R702))
*  Add comments and empty lines to improve the readability and maintainability of the control flow graph logic and the reaching definition analysis and live variable analysis functions ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR797-R817), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR842), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fL711-R854), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR900-R906), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR944), [link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-837b90142d1730f6a3ab20c91f1f35c95335ef82a021c74fd4dbdb05ff0e164fR1053-R1063))
*  Comment out a block of code in `compile_to_offloads` that was used for debugging purposes and restore the original functionality ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-8fde186587db97b3bbc8a856e59bc4467b30257335b0fad064b4eebd521a912bL87-R103))
*  Add a test case to `test_memory.py` to test the memory reuse feature for CUDA devices ([link](https://github.com/taichi-dev/taichi/pull/8065/files?diff=unified&w=0#diff-b506af0b07ab58c9384fedcbe8745c8b6eac88472fb60943a4d83f90d323bd43R49-R82))

### CODE REVIEW

File: taichi/ir/control_flow_graph.cpp [hunk 1]
```diff
@@ -492,67 +538,112 @@ void CFGNode::live_variable_analysis(bool after_lower_access) {
...
+  // dead-store-elimination strategy
+  //
+  // Here we use CFGNode::UseDefineStatus to mark whether a TensorType'd address
+  // is fully or partially modified.
+  std::unordered_map<Stmt *, CFGNode::UseDefineStatus> live_in_this_node;
+  std::unordered_map<Stmt *, CFGNode::UseDefineStatus> killed_in_this_node;
+
+  // Search for aliased addresses
+  // tensor_to_matrix_ptrs_map: map MatrixPtrStmt->origin to list of
+  // MatrixPtrStmts matrix_ptr_to_tensor_map: map MatrixPtrStmt to
```
Comment (dream189free) [inline taichi/ir/control_flow_graph.cpp:620]:
```suggestion
  //   MatrixPtrStmts 
  // matrix_ptr_to_tensor_map: map MatrixPtrStmt to
```

File: taichi/ir/control_flow_graph.cpp [hunk 2]
```diff
@@ -492,67 +538,112 @@ void CFGNode::live_variable_analysis(bool after_lower_access) {
...
+  // MatrixPtrStmt->origin
```
Comment (dream189free) [inline taichi/ir/control_flow_graph.cpp:621]:
```suggestion
  //   MatrixPtrStmt->origin
```

File: taichi/ir/control_flow_graph.cpp [hunk 3]
```diff
@@ -492,67 +538,112 @@ void CFGNode::live_variable_analysis(bool after_lower_access) {
...
+  std::unordered_map<Stmt *, std::vector<Stmt *>> tensor_to_matrix_ptrs_map;
+  std::unordered_map<Stmt *, Stmt *> matrix_ptr_to_tensor_map;
+  for (int i = begin_location; i < end_location; i++) {
+    auto stmt = block->statements[i].get();
+    if (stmt->is<MatrixPtrStmt>()) {
+      if (tensor_to_matrix_ptrs_map.count(stmt->as<MatrixPtrStmt>()->origin) ==
```
Comment (dream189free) [inline taichi/ir/control_flow_graph.cpp:628]: Consider giving a name to `stmt->as<MatrixPtrStmt>()->origin` for readability.

```
    auto stmt = block->statements[i].get();
    if (stmt->is<MatrixPtrStmt>()) {
      auto origin = stmt->as<MatrixPtrStmt>()->origin;
      if (tensor_to_matrix_ptrs_map.count(origin) == 0) {
        tensor_to_matrix_ptrs_map[origin] = {stmt};
      } else {
        tensor_to_matrix_ptrs_map[origin].push_back(stmt);
      }

      matrix_ptr_to_tensor_map[stmt] = origin;
    }
```

**Review by dream189free — APPROVED** [review]
LGTM!

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

**Pull request merged**